### ISSUE TITLE
Reducing ip camera RTMP latency

### ISSUE THREAD

**Issue by leodays** [CMU-Perceptual-Computing-Lab/openpose#447]
Are there any suggested methods for improving the speed at which openpose reads from an ip_camera stream?
I tested an existing RTMP stream using various clients (e.g. vlc, wowza, and http://support.akamai.com/flash/) and can get an almost real-time. When that same stream is used with the --ip_camera in openpose.bin the delay is nearly 15 seconds. I took a look at ipCameraReader.cpp and saw that it just uses the standard opencv:VideoCapture under the hood which seems to be the bottleneck. Any recommendations for making this faster?

**Comment by leodays** [comment]
I think it has to do the with VideoCapture buffer. I implemented a hack in the ipCameraReader.cpp:getRawFrame() method to verify and the video did get faster.
```
cv::Mat frame;
for (int i = 0; i < 6; i++) {
    frame = VideoCaptureReader::getRawFrame();
}
return frame;
```
If others have a better solution that would be great.

**Comment by gineshidalgo99** [comment]
I don't have a ip camera so i can't test it myself. But that's the same issue than for the webcam.

In the webcam, I solved it with a dedicated thread to continuously read frames (I'd say it's an improved version of @leodays ).

You could modify IPCameraReader with the idea in WebcamReader. Feel free to make a pull request if any of you implement it as it is in webcamreader (as I said, I can't try it myself that's why i didn't do it).

**Comment by afcentry** [comment]
请教一下，您调用的事本机摄像头还是通过IP调用的网络摄像头？网络摄像头的话是以怎么样的方式在哪个版本里面进行的？谢谢

**Comment by leodays** [comment]
@gineshidalgo99 I have a solution using pretty much the same steps in WebcamReader and it works. Sometimes however `openpose.bin` exits immediately with an error `Segmentation fault (core dumped)`. I'd like to submit a pull request once that is fixed and I can run openpose on an ip camera consistently .

Any idea why that might be happening? I used the same steps as WebcamReader, did you run into this problem for that approach?

**Comment by gineshidalgo99** [comment]
@leodays  That sounds as 2 threads trying to read/change the same variable (std::mutex not properly used), or maybe you copied cv::Mat by just doing cv::Mat b = a; (which doesn't copy, just share memory, so it might also crashes it), etc.

**Comment by leodays** [comment]
That's what I thought as well. The implementation is basically identical to webCamReader so I am using the std::mutex the same way. What's also interesting is that the few times it does start it runs perfectly reliably and does not crash. (If there was a multi-threading issue, I would expect that to crash eventually as well)

**Comment by gineshidalgo99** [comment]
You are right. Does it crash without your changes? Maybe it's the IP camera for some reason, not OpenPose nor your new code.

**Comment by leodays** [comment]
It does **not** crash **without** these new changes. The video is just very delayed (which we already know the root cause from the discussion above). Im sure it is something simple we are missing, if you/others have thoughts I'll try to implement the fix.

**Comment by leodays** [comment]
@gineshidalgo99 I think I found the root cause and solution.

I created a custom video stream to try to debug if this was an issue with the IP camera and learned that if the the stream has a high bit rate this issue occurs. This is probably due to some overflow happening in the swap when a frame is too large. (We could probably debug this more with a tool like Valgrind).

I don't have an absolute upper bound but I can confirm that streaming at 100 KB/s bitrate works perfectly and video quality is good as well. Ill submit a pull request in a few days with the changes. Thanks for you help.
